### ISSUE TITLE
FSDP FULL_SHARD Avoid Eviction for Final Layer

### ISSUE THREAD

**Issue by mvpatel2000** [pytorch/pytorch#115199]
### 🚀 The feature, motivation and pitch

Currently, with FSDP FULL_SHARD, the final layer is evicted and then all-gathered immediately after for the backward pass. This results in small bubbles switching between forward and backward (see attached trace). 

Ideally, we would optimize FSDP to avoid eviction. I can manually do this by tagging the last module as `SHARD_GRAD_OP`. It would be cool to automatically do this, either via FSDP flags or torch compile

CC: @awgu 

<img width="450" alt="image" src="https://github.com/pytorch/pytorch/assets/17102158/e875bf23-eac9-462b-9461-1e83e0a270dd">

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501

**Comment by awgu** [comment]
I agree with this concept. @kwen2501  had something for this, but we decided to focus on the new FSDP.

However, my current view as an FSDP developer is that I would prefer to not do this in eager FSDP. The reason is that right now, each FSDP instance can be responsible to free its own parameters. If we want to implement this feature, then I think we need to have a global system for managing the all-gathered parameters and only evicting/freeing them when necessary.

There is ongoing discussion on complexity in the FSDP design/implementation, and which parts we can offload to `torch.compile`. I would imagine that if we can trace through FSDP, then this is something that the compiler could do automatically. Of course, this needs more scoping to be sure though.

Would it help if we added a public API for setting an FSDP instance's sharding strategy (to `SHARD_GRAD_OP`)? In other words, is it inconvenient to make sure to pass `SHARD_GRAD_OP` to the FSDP constructor for just the last modules at a given module depth?

**Comment by mvpatel2000** [comment]
> my current view as an FSDP developer is that I would prefer to not do this in eager FSDP.

I agree with this -- this is much better suited for a compiler

> Would it help if we added a public API for setting an FSDP instance's sharding strategy?

Currently, we build a wrapper API around FSDP where users can specify wrapping structure by tagging modules with a constant `_fsdp_wrap` or a callable `fsdp_wrap_fn` which return `True/False/kwargs`. In the scenario of kwargs, we use `CustomPolicy`, which lets us override a FSDP instance's sharding strategy (eg to `SHARD_GRAD_OP`). You can see the code for this below.

https://github.com/mosaicml/composer/blob/5d20db1c345f2f06ed19658d769b004dbb4d320f/composer/trainer/dist_strategy.py#L448-L464

As long as `CustomPolicy` is supported in the rewrite (or something like that, which seems to be the case given you can explicitly wrap components individually if I understand the rewrite's design), I believe the existing API is sufficient
